1-perfect-hashing/perfect_hashing.cpp

```diff
@@ -0,0 +1,143 @@
+#include<bits/stdc++.h>
+
+using namespace std;
+
+typedef unsigned int uint;
+
+const int PRIME=41,MAX_TEAM_COUNT=37,MAX_PLAYER_COUNT=55;
+
+const INPUT_FILENAME="input.txt";
+
+struct Team
+{
+    string name;
+    vector<string> players;
+};
+
+int n;//count of teams
+Team teams[MAX_TEAM_COUNT];
+//vector with index i consists of strings with hash=i
+vector<string> hashed_players[MAX_PLAYER_COUNT],hashed_teams[MAX_TEAM_COUNT];
+
+void read()
+{
+    ifstream in(INPUT_FILENAME);
+	in>>n;
+	in.ignore();
+	for (int i=0;i<n;i++)
+    {
+        getline(in,teams[i].name);
+        int k;//count of players
+        in>>k;
+        in.ignore();
+        for (int j=0;j<k;j++)
+        {
+            string s;
+            getline(in,s);
+            teams[i].players.push_back(s);
+        }
+    }
+    in.close();
+}
+
+void print()
+{
+    for (int i=0;i<n;i++)
+    {
+        string s=teams[i].name;
+        cout<<s<<endl;
+        cout<<teams[i].players.size()<<endl;
+        for (int j=0;j<teams[i].players.size();j++)
+            cout<<teams[i].players[j]<<endl;
+    }
+}
+
+uint get_hash(string s)
+{
+    uint ans=0,pow=1;
+    for (int i=0;i<s.size();i++)
+    {
+        ans+=s[i]*pow;
+        pow*=PRIME;
+    }
+    return ans;
+}
+
+void hashing()
+{
+    for (int i=0;i<n;i++)
+    {
+        int ht=get_hash(teams[i].name)%MAX_TEAM_COUNT;//hash of the team's name
+        hashed_teams[ht].push_back(teams[i].name);
+        for (int j=0;j<teams[i].players.size();j++)
+        {
+            int hp=get_hash(teams[i].players[j])%MAX_PLAYER_COUNT;//hash of the player's name
+            hashed_players[hp].push_back(teams[i].players[j]);
+        }
+    }
+}
+
+void text_interface()
+{
+    char t=' ';
```

Comment:
Rename `t` to `action` or `command` etc.

Reply:
Done!